src/torch/gradchecks.py

```diff
@@ -0,0 +1,158 @@
+
+    torch_forward = SIRFTorchOperator(acq_model, image_data.clone())
+    torch_image_data = torch.tensor(image_data.as_array(), requires_grad=True).unsqueeze(0).cuda()
+
+    run_gradcheck(torch_forward, torch_image_data, (modality, data_type), "Forward",
+                  nondet_tol=1e-6, fast_mode=True, eps=1e-3, atol=1e-4, rtol=1e-4)
+
+
+@pytest.mark.skipif(not test_flags["adjoint"], reason="Adjoint test disabled")
+def test_adjoint_gradcheck(test_data):
+    acq_data, image_data, acq_model, modality, data_type = test_data
+    if modality == "PET":
+        pass
+    else:
+        pytest.skip("Tests not set up for other modalities at this time.")
+
+    adj_acq_model = sirf.SIRF.AdjointOperator(acq_model)
+    torch_adjoint = SIRFTorchOperator(adj_acq_model, acq_data.clone())
+    torch_acq_data = torch.tensor(acq_data.as_array(), requires_grad=True).unsqueeze(0).cuda()
+
+    run_gradcheck(torch_adjoint, torch_acq_data, (modality, data_type), "Adjointness",
+                  nondet_tol=1e-6, fast_mode=True, eps=1e-3, atol=1e-4, rtol=1e-4)
+
+
+
+@pytest.mark.skipif(not test_flags["objective_wrapped"], reason="Wrapped objective test disabled")
+def test_objective_function_with_wrapped_acquisition_model_gradcheck(test_data):
+    acq_data, image_data, acq_model, modality, data_type = test_data
+    if modality == "PET":
+        pass
+    else:
+        pytest.skip("Tests not set up for other modalities at this time.")
+
+    acq_model.set_background_term(acq_data.get_uniform_copy(1.))
+    acq_model.set_up(acq_data, image_data)
+
+    torch_forward = SIRFTorchOperator(acq_model, image_data.clone())
+    torch_image_data = torch.tensor(image_data.as_array(), requires_grad=True).unsqueeze(0).cuda()
+    loss = torch.nn.PoissonNLLLoss(log_input=False, full=False, reduction='sum')
+    torch_acq_data = sirf_to_torch(acq_data, device=torch_image_data.device)
+    torch_acq_model_obj = lambda x: loss(torch_forward(x), torch_acq_data)
+
+    run_gradcheck(torch_acq_model_obj, torch_image_data, (modality, data_type), "Objective (wrapped)",
+                  nondet_tol=1e-2, fast_mode=True, eps=1e-3, atol=1e-2, rtol=1e-2)
+
+
+@pytest.mark.skipif(not test_flags["objective"], reason="Objective test disabled")
+def test_objective_function_gradcheck(test_data):
+    acq_data, image_data, acq_model, modality, data_type = test_data
+    if modality == "PET":
+        obj_fun = pet.make_Poisson_loglikelihood(acq_data)
+    else:
+        pytest.skip("Tests not set up for other modalities at this time.")
+
+    obj_fun.set_acquisition_model(acq_model)
+    obj_fun.set_up(image_data)
+
+    torch_obj_fun = SIRFTorchObjectiveFunction(obj_fun, image_data.clone())
+    torch_image_data = torch.tensor(image_data.as_array(), requires_grad=True).unsqueeze(0).cuda()
+
+
+    run_gradcheck(torch_obj_fun, torch_image_data, (modality, data_type), "Objective",
+                  nondet_tol=1e-1, fast_mode=True, eps=1e-3, atol=1e-1, rtol=1e-1)
```

Comment:
@gschramm I had to set up the tolerance for non-determinism to be extremely high... I think this may be troubles with computing the numerical Jacobian from a PNLL type function causes issues with numerical precision? I was wondering if you get this with your gradcheck in parallelproj. As a comparison I only wrapped the acquisition model and used torch's PNLL for the objective, and I still need rather high tolerances for non-deteminism.